core/lib/zksync_core/src/state_keeper/io/seal_logic.rs

```diff
@@ -272,45 +271,61 @@ impl MiniblockSealCommand {
         self.seal_inner(storage, false).await;
     }
 
-    /// Seals a miniblock with the given number.
-    ///
-    /// If `is_fictive` flag is set to true, then it is assumed that we should seal a fictive miniblock
-    /// with no transactions in it. It is needed because there might be some storage logs / events
-    /// that are created after the last processed tx in the L1 batch: after the last transaction is processed,
-    /// the bootloader enters the "tip" phase in which it can still generate events (e.g.,
-    /// one for sending fees to the operator).
-    ///
-    /// `l2_erc20_bridge_addr` is required to extract the information on newly added tokens.
-    async fn seal_inner(&self, storage: &mut StorageProcessor<'_>, is_fictive: bool) {
-        self.assert_valid_miniblock(is_fictive);
-
-        let mut transaction = storage.start_transaction().await.unwrap();
-        if self.pre_insert_txs {
-            let progress = MINIBLOCK_METRICS.start(MiniblockSealStage::PreInsertTxs, is_fictive);
-            for tx in &self.miniblock.executed_transactions {
-                if let Ok(l1_tx) = L1Tx::try_from(tx.transaction.clone()) {
+    async fn insert_transactions(&self, transaction: &mut StorageProcessor<'_>) {
+        for tx_result in &self.miniblock.executed_transactions {
+            let mut tx = tx_result.transaction.clone();
+            // Override the "received at" timestamp for the transaction so that they are causally ordered (i.e., transactions
+            // with an earlier timestamp are persisted earlier). Without this property, code relying on causal ordering may work incorrectly;
+            // e.g., `pendingTransactions` subscriptions notifier can skip transactions.
+            tx.received_timestamp_ms = unix_timestamp_ms();
```

Comment:
Feels a little uncomfortable to set it here - the transaction was actually _received_ earlier.. But I understand the reasonning and cannot propose a better solution, so let's keep it

Reply:
We'll likely need to change this logic anyway when we have a different/more sophisticated mempool, so I'm fine both with the current solution and DAL. I personally prefer DAL, so if it doesn't add additional hustle let's change. As for `created_at` - yeah, let's not use it in app logic and keep just it only for journaling/debug

Reply:
JFYI, I have moved the timestamp assignment earlier into the transaction lifecycle, when a `Transaction` is converted into a `SyncAction`.